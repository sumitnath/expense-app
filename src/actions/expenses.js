import database from '../firebase/firebase';
import uuid from 'uuid';


export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense: expense
});

export const startAddExpense = (expenseData = {}) => {
    return (dispatch) => {
        const { description = '', note = '', amount = 0, createdAt = 0 } = expenseData;
        const expense = {
            description: description,
            note: note,
            amount: amount,
            createdAt: createdAt
        };
        return database.ref('expenses').push(expense)
            .then((ref) => {
                dispatch(addExpense({
                    id: ref.key,
                    ...expense
                }));
            });
    };
};

//Remove expense
export const removeExpense = ({ id } = {}) => ({
    type: 'REMOVE_EXPENSE',
    id: id
})

export const startRemoveExpense = ({ id } = {}) => {
    return (dispatch) => {
        return database.ref(`expenses/${id}`).remove()
            .then(() => {
                dispatch(removeExpense({ id }));
            });
    }
}

// Edit Expense
export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
})

export const startEditExpense = (id, updates) => {
    return (dispatch) => {
        return database.ref(`expenses/${id}`).update(updates)
            .then(() => {
                dispatch(editExpense(id, updates));
            })
    }
}

// SET_Expense
export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
});
// export const startSetExpenses;
export const startSetExpenses = () => {
    return (dispatch) => {
        return database.ref('expenses')
            .once('value')
            .then((snapshot) => {
                const expenses = [];
                snapshot.forEach(childSnapshot => {
                    expenses.push({
                        id: childSnapshot.key,
                        ...childSnapshot.val()
                    });
                });

                dispatch(setExpenses(expenses));
            });
    }
}