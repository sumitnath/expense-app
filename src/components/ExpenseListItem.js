import React from "react";
import { Link } from "react-router-dom";
//import { connect } from 'react-redux';
// we donot need the redux store as we have dont have remove here, so we remove connect()
// challenge remove remove button from  the ExpenseListItem.js and add it into EditExpensePage.js


const ExpenseListItem = ({ id, description, amount, createdAt }) => (

        <
        div >
        <
        Link to = { `/edit/${id}` } >
        <
        h3 > { description } < /h3> <
        /Link> <
        p > { amount }--{ createdAt } < /p>


        <
        /div>

    )
    //export default connect()(ExpenseListItem)
export default ExpenseListItem