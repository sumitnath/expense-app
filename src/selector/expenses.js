import moment from "moment"
// get visible data
export default (expenses, { text, sortBy, startDate, endDate }) => {
    return expenses.filter((expense) => {
        const createAtMoment = moment(expense.createdAt)
            // const startDateMatch = typeof startDate !== 'number' || expense.createdAt >= startDate
        const startDateMatch = startDate ? startDate.isSameOrBefore(createAtMoment, 'day') : true;
        // const endDateMatch = typeof endDate !== 'number' || expense.createdAt <= endDate
        const endDateMatch = endDate ? endDate.isSameOrAfter(createAtMoment, 'day') : true;
        const textMatch = expense.description.toLowerCase().includes(text.toLowerCase())
        return startDateMatch && endDateMatch && textMatch
    }).sort((a, b) => {
        if (sortBy === 'date') {
            return a.createdAt < b.createdAt ? 1 : -1
        } else if (sortBy === 'amount') {
            return a.amount < b.amount ? 1 : -1
        }
    })
}