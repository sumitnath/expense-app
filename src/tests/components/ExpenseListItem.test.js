import React from "react";
import { shallow } from "enzyme";
import ExpenseListItem from "../../components/ExpenseListItem";
import expenses from "../fixtures/expenses";
import toJSON from "enzyme-to-json";

test('Should render ExpenseList Item correctly', () => {
const wrapper = shallow(<ExpenseListItem {...expenses[0]} />)
expect(toJSON(wrapper)).toMatchSnapshot()
 })

 test('Should render ExpenseList Item without a Item', () => {
    const wrapper = shallow(<ExpenseListItem {...expenses} />)
    expect(toJSON(wrapper)).toMatchSnapshot()
     })