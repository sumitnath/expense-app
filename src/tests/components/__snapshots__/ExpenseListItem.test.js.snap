// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render ExpenseList Item correctly 1`] = `
<div>
  <Link
    to="/edit/1"
  >
    <h3>
       
      Gum
       
    </h3>
     
  </Link>
   
  <p>
     
    12536
    --
    0
     
  </p>
</div>
`;

exports[`Should render ExpenseList Item without a Item 1`] = `
<div>
  <Link
    to="/edit/undefined"
  >
    <h3>
       
       
    </h3>
     
  </Link>
   
  <p>
     
    --
     
  </p>
</div>
`;
